qiskit_experiments/calibration_management/calibrations.py

```diff
@@ -1099,63 +1077,58 @@ def get_schedule(
-        instruction and the qubits that are in said subroutine. This requires a
-        careful extraction of the qubits from the subroutine and in the appropriate
-        order. Next, the parameters are identified and assigned. This is needed to
-        handle situations where the same parameterized schedule is called but on
-        different channels. For example,
-
-        .. code-block:: python
-
-            ch0 = Parameter("ch0")
-            ch1 = Parameter("ch1")
-
-            with pulse.build(name="xp") as xp:
-                pulse.play(Gaussian(duration, amp, sigma), DriveChannel(ch0))
-
-            with pulse.build(name="xt_xp") as xt:
-                pulse.call(xp)
-                pulse.call(xp, value_dict={ch0: ch1})
-
-        Here, we define the xp :class:`ScheduleBlock` for all qubits as a Gaussian. Next, we define
-        a schedule where both xp schedules are called simultaneously on different channels. We now
-        explain a subtlety related to manually assigning values in the case above. In the schedule
-        above, the parameters of the Gaussian pulses are coupled, e.g. the xp pulse on ch0 and ch1
-        share the same instance of :class:`ParameterExpression`. Suppose now that both pulses have
-        a duration and sigma of 160 and 40 samples, respectively, and that the amplitudes are 0.5
-        and 0.3 for qubits 0 and 2, respectively. These values are stored in self._params. When
-        retrieving a schedule without specifying assign_params, i.e.
-
-        .. code-block:: python
-
-            cals.get_schedule("xt_xp", (0, 2))
+        """Recursively removes all called by name instructions."""
+        ret_schedule = ScheduleBlock(
+            name=schedule.name, alignment_context=schedule.alignment_context
+        )
 
-        we will obtain the expected schedule with amplitudes 0.5 and 0.3. When specifying the
-        following :code:`assign_params = {("amp", (0,), "xp"): Parameter("my_new_amp")}` we
-        will obtain a schedule where the amplitudes of the xp pulse on qubit 0 is set to
-        :code:`Parameter("my_new_amp")`. The amplitude of the xp pulse on qubit 2 is set to
-        the value stored by the calibrations, i.e. 0.3.
+        for block in schedule.blocks:
+            if isinstance(block, CalledScheduleByName):
+
+                # Down-select the qubits to those of the called instruction.
+                called_qubits = []
+                for ch in block.channels:
+                    # Ignore ControlChannels as their index does not map to a qubit index.
+                    if not isinstance(ch, ControlChannel):
+                        called_qubits.append(self._get_channel_index(qubits, ch))
+
+                    called_qubits = tuple(called_qubits)
+
+                # If no qubits were down-selected then use all the qubits.
+                if len(called_qubits) == 0:
+                    called_qubits = qubits
```

Comment:
What is the need for this? Users should be specifying the necessary channels for `CalledScheduleByName`.

Reply:
There might be some room for improvement in the design here. Let me first flesh out why this is here. Consider the case as in `TestCoupledAssigning` in `test_calibrations.py`. We have a schedule
```python
with pulse.build(name="cr_p") as cr_p:
    pulse.play(GaussianSquare(dur, amp_cr, sigma, width), c1_)  # c1_ is a control channel.
```
Next, we build a schedule on top of this by calling `cr_p` by name, e.g., 
```python
ecr = pulse.ScheduleBlock(name="ecr", alignment_context=AlignSequential())
ecr.append(CalledScheduleByName("cr_p", c1_))
...
```
Notice already that `cr_p` does not know the qubits, only the control channel. Now, in this test we are working with qubits (3, 2) the associated control channel is 10. Therefore, if I do `called_qubits.append(self._get_channel_index(qubits, ch))` as done a few lines above the code you point out, the  I do not get any qubits back because `cr_p` does not know the qubits.

Reply:
As a start, this should already be better: [1c17c83](https://github.com/Qiskit/qiskit-experiments/pull/725/commits/1c17c837183f2f032ad16cfd46ee6b62ec035bbe) . However, I think this logic appears again in another location of the code and it might be wise to outsource it to a separate function.

Reply:
I was thinking of the channels in `CalledScheduleByName` as playing a different role. What is really needed is a `qubits` label, but because we support templating schedules and filling in the qubit indices later we need that label to support parametrized channels. So I was thinking `channels` would be likely just a set of `DriveChannel`s with the indices possibly parameters like `Parameter("ch0")` even if the referenced schedule actually used a `ControlChannel` instead of `DriveChannel`s.

So for my way of thinking about it, I would reverse the order of the `_resolve_called_by_name` call and the `assign_parameters` for the channels in `get_schedule`. Then all the channels are given integer indices before `_resolve_called_by_name` and the qubits for `get_schedule()` should just be `qubits = tuple(c.index for c in block.channels)`.

Maybe this is deviating too far from how channels are really used? Instead of channels, we could take a `qubits` argument directly and allow either `int` or `Parameter("ch0")`-like entries.